packages/core/src/html/UlNodeProcessor.ts

```diff
@@ -0,0 +1,256 @@
+    size: defaultIcon?.size,
+    className: defaultIcon?.className,
+  };
+
+  // Overwrite defaults with the actual attributes from node if they exist
+  return {
+    ...iconAttributes,
+    icon: node.attribs.icon !== undefined ? node.attribs.icon : iconAttributes.icon,
+    width: node.attribs.width !== undefined ? node.attribs.width : iconAttributes.width,
+    height: node.attribs.height !== undefined ? node.attribs.height : iconAttributes.height,
+    size: node.attribs.size !== undefined ? node.attribs.size : iconAttributes.size,
+    className: node.attribs.class !== undefined ? node.attribs.class : iconAttributes.className,
+  };
+};
+
+// Helper function to update UL node
+const updateUlNode = (node: MbNode, icon: IconAttributes) => {
+  if (node.attribs) {
+    node.attribs.icon = icon.icon;
+    node.attribs.width = icon.width;
+    node.attribs.height = icon.height;
+    node.attribs.size = icon.size;
+    node.attribs.class = icon.className;
+  }
+};
+
+function updateLiChildren(child: NodeOrText, defaultLiIconAttributes: IconAttributes) {
+  const childNode = child as MbNode;
+
+  const curLiIcon = getIconAttributes(childNode, defaultLiIconAttributes);
+
+  ['icon', 'width', 'height', 'size', 'class'].forEach((attr) => {
+    delete childNode.attribs[attr];
+  });
+
+  const children = cheerio(child.children);
+
+  // Create a new div and span
+  const div = cheerio('<div></div>');
+  const iconSpan = createIconSpan(curLiIcon!);
+
+  // Append each child to the div
+  children.each((index, elem) => {
+    div.append(cheerio(elem));
+  });
+
+  // Empty the child
+  cheerio(child).empty();
+
+  // Append iconSpan and div to the child
+  cheerio(child).append(iconSpan);
+  cheerio(child).append(div);
+
+  cheerio(child).css({
+    display: 'flex',
+  });
+}
+
+function isRelevantNode(node: NodeOrText): boolean {
+  return 'name' in node && node.name === 'ul' && Boolean(node.children);
+}
+
+function updateFirstUl(ulNode: MbNode): IconAttributes | null {
+  let isFirstUlLiChild = true;
+  let defaultIcon: IconAttributes | null = null;
+
+  for (let i = 0; i < ulNode.children.length; i += 1) {
+    const ulChild = ulNode.children[i];
+    if (ulChild.name === 'li') {
+      if (isFirstUlLiChild) {
+        const iconAttr = getIconAttributes(ulChild as MbNode);
+        if (iconAttr) {
+          defaultIcon = iconAttr;
+          isFirstUlLiChild = false;
+        } else return null;
+      } else if (defaultIcon) {
+        defaultIcon = getIconAttributes(ulChild as MbNode, defaultIcon);
+      }
+    }
+  }
+
+  return defaultIcon;
+}
+
+function updateRestUl(ulNode: MbNode, defaultIcon: IconAttributes): IconAttributes {
+  updateUlNode(ulNode, defaultIcon);
+
+  const updatedIcon = ulNode.children
+    .filter(ulChild => ulChild.name === 'li')
+    .reduce((icon, ulChild) => getIconAttributes(ulChild as MbNode, icon)!, defaultIcon);
+
+  return updatedIcon;
+}
+
+export function waterfallModel(node: NodeOrText): NodeOrText {
```

Comment:
3rd/4th/etc. nested levels seems to have some issues, let's see if we can find a way to do this recursively.

![image](https://github.com/MarkBind/markbind/assets/3306138/59452a2e-13a8-47b3-99a3-7481aa23d6b3)